src/core/Akka.Persistence/Serialization/PersistenceMessageSerializer.cs

```diff
@@ -148,18 +169,54 @@ private PersistentFSMSnapshot GetPersistentFSMSnapshot(object obj)
             return message;
         }
 
-        public override object FromBinary(byte[] bytes, Type type)
+        public override object FromBinary(byte[] bytes, string manifest)
         {
-            if(type == null) return GetPersistentRepresentation(PersistentMessage.Parser.ParseFrom(bytes));
-            if (type == typeof(Persistent)) return GetPersistentRepresentation(PersistentMessage.Parser.ParseFrom(bytes));
-            if (type == typeof(IPersistentRepresentation)) return GetPersistentRepresentation(PersistentMessage.Parser.ParseFrom(bytes));
-            if (type == typeof(AtomicWrite)) return GetAtomicWrite(bytes);
-            if (type == typeof(AtLeastOnceDeliverySnapshot)) return GetAtLeastOnceDeliverySnapshot(bytes);
-            if (type == typeof(PersistentFSM.StateChangeEvent)) return GetStateChangeEvent(bytes);
-            if (type.GetTypeInfo().IsGenericType
-                && type.GetGenericTypeDefinition() == typeof(PersistentFSM.PersistentFSMSnapshot<>)) return GetPersistentFSMSnapshot(type, bytes);
+            switch (manifest)
+            {
+                case null:
+                case IPersistentRepresentationManifest:
+                case IPersistentRepresentationManifestNetCoreManifest:
+                    return GetPersistentRepresentation(PersistentMessage.Parser.ParseFrom(bytes));
+
+                case AtomicWriteManifest:
+                case AtomicWriteNetCoreManifest:
+                    return GetAtomicWrite(bytes);
+
+                case AtLeastOnceDeliverySnapshotManifest:
+                case AtLeastOnceDeliverySnapshotNetCoreManifest:
+                    return GetAtLeastOnceDeliverySnapshot(bytes);
+
+                case StateChangeEventManifest:
+                case StateChangeEventNetCoreManifest:
+                    return GetStateChangeEvent(bytes);
+
+                case PersistentFSMSnapshotManifest:
+                case var m when m.StartsWith(PersistentFSMSnapshotNetCoreManifest):
+                    return GetPersistentFSMSnapshot(bytes);
+            }
 
-            throw new SerializationException($"Unimplemented deserialization of message with type [{type}] in [{GetType()}]");
+            throw new SerializationException($"Unimplemented deserialization of message with manifest [{manifest}] in [{GetType()}]");
+        }
+
+        public override string Manifest(object obj)
```

Comment:
For the time being, we should return the legacy types - I believe there are some persistence plugins (i.e. SQL ones) that are still relying on manifest data using reflection and while I would like to eventually clean that up, probably best to change as little as possible here first.